src/Processors/QueryPlan/Optimizations/optimizeReadInOrder.cpp

```diff
@@ -0,0 +1,934 @@
+                            fixed_columns.insert(frame.node);
+                    }
+                    else if (frame.node->type == ActionsDAG::ActionType::FUNCTION)
+                    {
+                        if (frame.node->function_base->isDeterministicInScopeOfQuery())
+                        {
+                            //std::cerr << "*** enreachFixedColumns check " << frame.node->result_name << std::endl;
+                            bool all_args_fixed_or_const = true;
+                            for (const auto * child : frame.node->children)
+                            {
+                                if (!child->column && !fixed_columns.contains(child))
+                                {
+                                    //std::cerr << "*** enreachFixedColumns fail " << child->result_name <<  ' ' << static_cast<const void *>(child) << std::endl;
+                                    all_args_fixed_or_const = false;
+                                }
+                            }
+
+                            if (all_args_fixed_or_const)
+                            {
+                                //std::cerr << "*** enreachFixedColumns add " << frame.node->result_name << ' ' << static_cast<const void *>(frame.node) << std::endl;
+                                fixed_columns.insert(frame.node);
+                            }
+                        }
+                    }
+                }
+
+                stack.pop();
+            }
+        }
+    }
+}
+
+/// This structure stores a node mapping from one DAG to another.
+/// The rule is following:
+/// * Input nodes are mapped by name.
+/// * Function is mapped to function if all children are mapped and function names are same.
+/// * Alias is mapped to it's children mapping.
+/// * Monotonic function can be mapped to it's children mapping if dirrect mapping does not exist.
+///   In this case, information about monotonicity is filled.
+/// * Mapped node is nullptr if there is no mapping found.
+///
+/// Overall, directly mapped nodes represent equal calculations.
+/// Notes:
+/// * Mapped DAG can contain many nodes which represent the same calculation.
+///   In this case mapping is ambiguous and only one node is mapped.
+/// * Aliases for mapped DAG are not supported.
+/// DAG for PK does not contain aliases and ambiguous nodes.
+struct MatchedTrees
+{
+    /// Monotonicity is calculated for monotonic functions chain.
+    /// Chain is not strict if there is any non-strict monotonic function.
+    struct Monotonicity
+    {
+        int direction = 1;
+        bool strict = true;
+    };
+
+    struct Match
+    {
+        const ActionsDAG::Node * node = nullptr;
+        std::optional<Monotonicity> monotonicity;
+    };
+
+    using Matches = std::unordered_map<const ActionsDAG::Node *, Match>;
+};
+
+MatchedTrees::Matches matchTrees(const ActionsDAG & inner_dag, const ActionsDAG & outer_dag)
```

Comment:
This function is generally well written, we need to just add some internal comments, for each step that follows documentation above.

Reply:
I consider moving it somewhere to common interface, but likely will do it later (after aggregation-in-order)